config/v1/types_infrastructure.go

```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
 	Template string `json:"template,omitempty"`
 }
 
+// VSphereFailureDomainAffinity contains the vCenter cluster vm-host group (virtual machine and host types)
+// and the vm-host affinity rule that together creates a affinity configuration for vm-host based zonal.
+// This configuration within vCenter creates the required association between a failure domain, virtual machines
+// and ESXi hosts to create a vm-host based zone.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup) : true",message="when type is HostGroup, hostGroup field must be defined"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type != 'HostGroup' ?  !has(self.hostGroup) : true",message="when hostGroup field is defined, type must be HostGroup"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmGroup) : true",message="when type is HostGroup, vmGroup field must be defined"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type != 'HostGroup' ?  !has(self.vmGroup) : true",message="when vmGroup field is defined, type must be HostGroup"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmHostRule) : true",message="when type is HostGroup, vmHostRule field must be defined"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type != 'HostGroup' ?  !has(self.vmHostRule) : true",message="when vmHostRule field is defined, type must be HostGroup"
+// +union
+type VSphereFailureDomainAffinity struct {
+	// +kubebuilder:validation:Optional
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=HostGroup;ComputeCluster;Datacenter
+	// +unionDiscriminator
+	Type VSphereFailureDomainType `json:"type"`
+	// vmGroup is the name of the vm-host group of type virtual machine within vCenter for this failure domain.
+	// This field is required when the VSphereFailureDomain ZoneType is HostGroup
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:MaxLength=80
+	// +optional
+	// +unionMember,optional
+	VMGroup string `json:"vmGroup,omitempty"`
+
+	// hostGroup is the name of the vm-host group of type host within vCenter for this failure domain.
+	// This field is required when the VSphereFailureDomain ZoneType is HostGroup
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:MaxLength=80
+	// +optional
+	// +unionMember,optional
```

Comment:
You mean `+unionMember`, the `optional` here means that you can specify the type HostGroup without supplying this field, it's an optional bit of information to add to the context, but not required.

You've validated above that the field is required

```suggestion
	// +unionMember
```

Reply:
When setting `RegionAffinity` only the type is required, all the string fields of `VSphereFailureDomainAffinity` are not required.

Reply:
Re-using this type between the region and zone affinities is likely to cause headache if they have validation requirements that are different to one another